### ISSUE TITLE
Avoid mutating the arguments to class constructors and .set

### ISSUE THREAD

**Issue by marbemac** [jagi/meteor-astronomy#429]
``` js
const myPlainObject = {foo: 'bar'};
model.set(myPlainObject);
```

The set operation seems to mutate the passed in argument, which can lead to subtle and difficult to trace bugs (which we just ran in to). In particular, the one we ran into has to do with the typing of nested classes. If myPlainObject has a plain property that is defined as a nested class in model, that property will be mutated, and turned into an instance of that class (which is what we want on model, but not on the original object that is passed in).

Ideally, one could assume no side effects on the variables and objects that I pass into astronomy methods like set, which means not mutating those arguments. For now, we're cloning objects before passing them into .set.

Does this make sense? Thoughts?

**Comment by lukejagodzinski** [comment]
Oh I see. But I don't think that it's always expected behavior. Maybe you will have a case where having field of the `Object` type, you assign a reference to it. Why? Because you may later need to compare (`===`) this value with other reference.

**Comment by marbemac** [comment]
I guess. IMHO the types of bugs this can produce are pretty bad since they are very hard to trace, and manifest themselves subtly (as most mutation related bugs are). The use case you're describing seems like an edge case.

Right now we're going through and adding cloneDeep's everywhere we pass plain objects into astronomy constructors, to avoid the unintended side-effects we've been running into.

**Comment by lukejagodzinski** [comment]
I was thinking about that and in deed objects passed to constructor or to set method should be cloned. Developer will still be able to assign reference by direct assignment:

``` js
doc.fieldName = refObj;
```

**Comment by lukejagodzinski** [comment]
I'm just a little bit worried about performance when cloning documents. The best way to clone document is using `EJSON.clone` which is a little bit slow. The only situation when argument passed to constructor should be cloned is when developer is instantiating a document instance by him/herself.

``` js
var post = new Post({field: 'value'});
```

It shouldn't happen when document is being fetched from the collection because in such situation we don't need to do it, so it would be unnecessary waste of performance.

``` js
var post = Post.findOne();
```

In Astronomy 1.x I was cloning objects in all possible cases to make sure that references are not passed and original objects are not modified but that came with the huge cost. So I need to make sure that it will be performant.

**Comment by marbemac** [comment]
Totally get the performance concern! Here we are mainly talking about the manual instantiation issue, where the user is passing in an object they are likely managing somewhere else (in a local cache like redux for example), and would not expect that object to be mutated (and thus, their local cache to be mutated). Would it be possible to clone on manual instantiation and .set? This seems like the best compromise between performance and UX.

Anyways, up to you obviously. This is an issue that has bitten us in our large production application several times, in hard to figure out ways.

**Comment by lukejagodzinski** [comment]
Yes it's possible. In 2.1 I've prepared the mechanism that will allow that. I will make that fix in 2.1.1

**Comment by marbemac** [comment]
Thank you :)
